Set timestamps on `insert_all`/`upsert_all` record creation

sambostock opened the pull request:
### Summary

<!-- Provide a general description of the code changes in your pull
request... were there any bugs you had fixed? If so, mention them. If
these bugs have open GitHub issues, be sure to tag them here as well,
to keep the conversation linked together. -->

This provides the option to set timestamp columns (`created_at`, `created_on`, `updated_at`, `updated_on`) automatically when using `insert_all` or `upsert_all` (and related methods) results in a created record. Currently, the only clean way to ensure they are set is to either set a default on the columns themselves, or to provide them explicitly as attributes while also customizing the `on_duplicate` SQL so as not to overwrite `created_at` on existing records.

Builds on #42993.

### Other Information

This is still a work in progress and needs some cleanup (`TODO`s, etc.) but I think is ready for preliminary review on the approach.

<!-- If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

If you are updating any of the CHANGELOG files or are asked to update the
CHANGELOG files by reviewers, please add the CHANGELOG entry at the top of the file.

Finally, if your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

Thanks for contributing to Rails! -->

## ⚠️ Before Merging
- [x] Rebase on `main` after #42993 has been merged
- [ ] Add CHANGELOG entry
- [ ] Update documentation (and guides?)
- [ ] Address TODO comments and open questions.

sambostock commented on activerecord/lib/active_record/insert_all.rb:
```diff
@@ -7,11 +7,12 @@ class InsertAll # :nodoc:
     attr_reader :on_duplicate, :returning, :unique_by, :update_sql
 
-    def initialize(model, inserts, on_duplicate:, returning: nil, unique_by: nil)
+    def initialize(model, inserts, on_duplicate:, returning: nil, unique_by: nil, set_timestamps_on_create: false)
```
I made this behaviour configurable, as I figured some people might want to opt-out for whatever reasons, similarly to being able to opt-out of setting `updated_at` and `updated_on` automatically on upserts.

I made it default to `false` because I wasn't sure if we'd want to default to `true` right off the bat. That said, with Rails 7 coming up, maybe we could? 🤷 

I think ideally, we'd either ship with it on by default, or set it to on by default in a subsequent version of ActiveRecord.

sambostock commented on activerecord/lib/active_record/insert_all.rb:
```diff
@@ -64,15 +65,25 @@ def map_key_with_value
       end
     end
 
+    # TODO: Consider remaining this method, as it only conditionally extends keys, not always
+    def keys_including_timestamps
+      @keys_including_timestamps ||= if set_timestamps_on_create?
+        keys + model.all_timestamp_attributes_in_model
+      else
+        keys
+      end
+    end
```
I don't like this name, but I'm not sure of a better alternative. `extended_keys` maybe, but that's not super clear about how they're extended.

sambostock commented on activerecord/lib/active_record/insert_all.rb:
```diff
@@ -148,10 +159,18 @@ def disallow_raw_sql!(value)
+      end
+
+      def timestamps_for_create
+        model.all_timestamp_attributes_in_model.index_with(connection.high_precision_current_timestamp)
```
`connection.high_precision_current_timestamp` was added in #42993

sambostock commented on activerecord/test/cases/insert_all_test.rb:
```diff
@@ -393,6 +393,66 @@ def test_upsert_all_uses_given_updated_on_over_implicit_updated_on
+    created_at = Time.now.utc - 30
+    Book.create! id: 101, name: "Out of the Silent Planet", published_on: Date.new(1938, 4, 1), created_at: created_at, updated_at: created_at, updated_on: created_at.to_date
+
+    Book.upsert_all [{ id: 101, name: "Out of the Silent Planet", published_on: Date.new(1938, 4, 1) }], set_timestamps_on_create: true
+
+    book = Book.find(101)
+
+    assert_in_delta created_at, book.created_at, 1
+    assert_equal book.created_at, book.updated_at
+    assert_equal book.updated_at.to_date, book.updated_on
+  end
+
+  def test_upsert_all_does_not_touch_creation_timestamps_on_updated_record_when_told_to_set_timestamps_on_create
+    # TODO: Should we skip this under any circumstances?
+
+    created_at = Time.now.utc - 30
+    Book.create! id: 101, name: "Out of the Silent Planet", published_on: Date.new(1938, 4, 1), created_at: created_at, updated_at: created_at, updated_on: created_at.to_date
+
+    Book.upsert_all [{ id: 101, name: "Out of the Silent Planet", published_on: Date.new(1938, 4, 8) }], set_timestamps_on_create: true
+
+    book = Book.find(101)
+
+    assert_in_delta created_at, book.created_at, 1
+    assert_not_equal book.created_at, book.updated_at
+    assert_equal book.updated_at.to_date, book.updated_on
+  end
+
+  def test_upsert_uses_explicit_timestamps_even_if_told_to_set_timestamps_on_create
+    # TODO: Should we skip this under any circumstances?
+
+    created_at = Time.now.utc - 30
+
+    Book.upsert_all [{ id: 101, name: "Out of the Silent Planet", published_on: Date.new(1938, 4, 8), created_at: created_at, updated_at: created_at, updated_on: created_at.to_date }], set_timestamps_on_create: true
+
+    book = Book.find(101)
+
+    assert_in_delta created_at, book.created_at, 1
+    assert_equal book.created_at, book.updated_at
+    assert_equal book.updated_at.to_date, book.updated_on
+  end
```
These tests all exercise the API through `upsert_all`. Is that sufficient, or should we test its usage in the various other methods (e.g. `insert_all`)?

tenderlove commented:
For code like `Book.upsert_all [{ id: 101, name: "Out of the Silent Planet", published_on: Date.new(1938, 4, 1) }],`, would it be possible to detect that doing something like `Book.create({ id: 101, name: "Out of the Silent Planet", published_on: Date.new(1938, 4, 1) })` *would have* added a `created_at`, and use that behavior as the default?  As opposed to specifying an option.

AFAIK `upsert_all` / `insert_all` are supposed to basically just be "performance hacks" as they don't require instantiating an AR model for each record and then inserting.  But it still seems to me that if we can mirror at least some of the behavior, then shouldn't we?

OTOH people could just be looking at `insert_all` / `update_all` as "I don't feel like writing a bunch of SQL, but I really want it to insert exactly what I specified".  I'm not 100% clear on the intention.

@jeremy do you have any thoughts on this?

sambostock commented:
Thanks for the review!

@tenderlove just to confirm, and if I'm reading [the `ActiveRecord::Timestamp` API docs](https://api.rubyonrails.org/classes/ActiveRecord/Timestamp.html) correctly, you're suggesting using `config.active_record.record_timestamps` as the default when deciding if we should set timestamps or not? I do like this as a default. Would we want it the option to override (either also as a config, or on a per-call/per-model basis)? I only really mention it because we allow controling the behaviour on update, so we probably want the APIs to be consistent.

jeremy commented:
At first blush, I would expect insert_all (and hence upsert_all) to set created_at. But maybe it shouldn't, if update_all is the analog to follow.

As a thought experiment, if the method were named something like create_many, we'd expect it to behave just like create. So the "insert" language screams "low-level, and only what's specified."

Given that, maybe 1. documentation and 2. a user-friendly way to manage column default values (not timestamp-specific) would be sufficient.

sambostock commented:
@jeremy what about the fact that we update timestamps on duplicates? As a user, it feels confusing that:

- by default, timestamps are magically set
- but when upserting, updated rows have their `updated_{at,on}` set, but created rows have none of their timestamps set

I feel like we should either manage all timestamp attributes in all circumstances, or none in any circumstances, while perhaps providing an option to control this behaviour. Given that `record_timestamps` exists on the model, we could use that as a default, while also allowing a `record_timestamps:` keyword argument to force the behaviour in one direction or the other for a single method call.

jeremy commented:
> we should either manage all timestamp attributes in all circumstances, or none in any circumstances, while perhaps providing an option to control this behaviour

Agreed.

> Given that record_timestamps exists on the model, we could use that as a default, while also allowing a record_timestamps: keyword argument to force the behaviour in one direction or the other for a single method call.

Seems reasonable to me!

sambostock committed:
Optimize loop in insert_all

We can skip looping over timestamp_attributes_for_update_in_model
entirely if the condition is false, instead of checking it for each
attribute.

sambostock committed:
Add tests for existing timestamp behaviour

These tests document the fact that upsert_all

- never sets creation timestamps
- always sets update timestamps, on update

regardless of record_timestamps config.

sambostock committed:
Respect model.record_timestamps? in upsert_all

Typically, Active Record respects the record_timestamps config on models
when deciding to automatically set timestamp columns:

- {created,updated}_{at,on} are set on record creation
- updated_{at,on} are set on record update

Currently, the machinery for bulk insert/upsert only automatically sets
timestamps

- on update, regardless of record_timestamps config

It does not set any timestamps on record creation.

This teaches it how to set timestamps on creation, and respect
record_timestamp, so the behavior matches typical record operations.

sambostock committed:
Allow record_timestamps override on {in,up}sert_all

This provides the option to override the record_timestamps value
provided by the model when bulk inserting/upserting records, which
controls if timestamp columns are set automatically.

sambostock committed:
Add high precision creation upsert regression test

This test ensures we use the same high precision timestamp for bulk
record creation as we do for bulk record updates.

See #42993 for details.

sambostock committed:
Add timestamp column alias test for upserts

This test ensures the bulk insert/upsert machinery for automatically
updating timestamps correctly handles tables where the column are
aliased.
